app/Commands/Dev/Tree/Compile/Base.hs

```diff
@@ -0,0 +1,93 @@
+module Commands.Dev.Tree.Compile.Base where
+
+import Commands.Base
+import Commands.Dev.Tree.Compile.Options
+import Commands.Extra.Compile qualified as Compile
+import Juvix.Compiler.Asm.Pretty qualified as Asm
+import Juvix.Compiler.Backend qualified as Backend
+import Juvix.Compiler.Backend.C qualified as C
+import Juvix.Compiler.Nockma.Pretty qualified as Nockma
+import Juvix.Compiler.Tree.Data.InfoTable qualified as Tree
+
+data PipelineArg = PipelineArg
+  { _pipelineArgOptions :: CompileOptions,
+    _pipelineArgFile :: Path Abs File,
+    _pipelineArgTable :: Tree.InfoTable
+  }
+
+getEntry :: (Members '[Embed IO, App, TaggedLock] r) => PipelineArg -> Sem r EntryPoint
+getEntry PipelineArg {..} = do
+  ep <- getEntryPoint (AppPath (preFileFromAbs _pipelineArgFile) True)
+  return $
+    ep
+      { _entryPointTarget = getTarget (_pipelineArgOptions ^. compileTarget),
+        _entryPointDebug = _pipelineArgOptions ^. compileDebug,
+        _entryPointUnsafe = _pipelineArgOptions ^. compileUnsafe,
+        _entryPointOptimizationLevel = fromMaybe defaultOptLevel (_pipelineArgOptions ^. compileOptimizationLevel),
+        _entryPointInliningDepth = _pipelineArgOptions ^. compileInliningDepth
+      }
+  where
+    getTarget :: CompileTarget -> Backend.Target
+    getTarget = \case
+      TargetWasm32Wasi -> Backend.TargetCWasm32Wasi
+      TargetNative64 -> Backend.TargetCNative64
+      TargetGeb -> Backend.TargetGeb
+      TargetVampIR -> Backend.TargetVampIR
+      TargetCore -> Backend.TargetCore
+      TargetAsm -> Backend.TargetAsm
+      TargetTree -> Backend.TargetTree
+      TargetNockma -> Backend.TargetNockma
+
+    defaultOptLevel :: Int
+    defaultOptLevel
+      | _pipelineArgOptions ^. compileDebug = 0
+      | otherwise = defaultOptimizationLevel
+
+runCPipeline ::
+  forall r.
+  (Members '[Embed IO, App, TaggedLock] r) =>
+  PipelineArg ->
+  Sem r ()
+runCPipeline pa@PipelineArg {..} = do
+  entryPoint <- getEntry pa
+  C.MiniCResult {..} <- getRight (run (runReader entryPoint (runError (treeToMiniC _pipelineArgTable :: Sem '[Error JuvixError, Reader EntryPoint] C.MiniCResult))))
+  cFile <- inputCFile _pipelineArgFile
+  embed @IO (writeFile (toFilePath cFile) _resultCCode)
```

Comment:
we recently added `writeFileEnsureLn` to the prelude. Should we use that normally? I'd say yes

Reply:
I will leave this for future refactor. There are 3 places just in this single file with `writeFile` used this way, and I think it's better if we change all of them at once, to be consistent.